Socket-activated client helper

julian-klode opened the pull request:
Introduce support for a socket-activated client helper for aptcc. This solves a problem where the current client helpers are children of the client starting the transaction and potentially bound to the client's lifetime; so if the client is exited early, while the transaction is still running, no debconf prompts would be shown.

A socket-activated helper solves this problem by making sure we only ever run one helper as a systemd user service.

This change is incompatible with the debconf-kde hacks, I'm not sure what to do about those yet.

julian-klode commented:
I think it would be wise to make the .socket unit remove the socket if it is stopped, so we don't end up in a situation where the socket exists but is not usable.

hughsie commented:
@robert-ancell -- can you review these please? Thanks.

dantti commented:
> This change is incompatible with the debconf-kde hacks, I'm not sure what to do about those yet.

If this breaks debconf-kde we need a fix before merging it.

julian-klode commented:
> > This change is incompatible with the debconf-kde hacks, I'm not sure what to do about those yet.
> 
> If this breaks debconf-kde we need a fix before merging it.

This prevents packagekit-glib2 from using debconf-kde. I'm not sure anybody is actually testing this. I think debconf-kde needs to gather the ability to launch with an open socket / socket activation, and then pk-debconf-helper could just exec to it if it does.

dantti commented:
alright, I thought the change would break it when running like it's done in Apper

julian-klode commented:
So, looking at this again after the Christmas break:

This change also breaks if we are running in a terminal, because we'd suddenly be talking to a systemd-activated frontend which does not have our tty. That said, IIRC, it did not work for me anyway, but I'll probably have to reverify.

We do have a fallback to the current way currently in case debconf-socket does not exist. We can easily extend that check to also use the current way if we are running on kde or in a terminal (`DEBIAN_FRONTEND={kde,dialog}` set by `pk_client_create_helper_argv_envp()`).

This also avoids any possible new breakage regarding KDE.

robert-ancell commented on configure.ac:
```diff
@@ -216,10 +216,16 @@ if test x$enable_systemd = xyes; then
+		    AS_HELP_STRING([--with-systemduserunitdir=DIR], [Directory for systemd user service files]),
+		    [has_systemduserunitdir=$with_systemduserunitdir],
+		    [has_systemduserunitdir=$($PKG_CONFIG --variable=systemduserunitdir systemd)])
+	AC_SUBST([systemduserunitdir], [$has_systemduserunitdir])
```
This line is duplicated.

robert-ancell commented on backends/aptcc/pk-debconf-helper.c:
```diff
@@ -0,0 +1,54 @@
+static GMainLoop *main_loop;
+static PkClientHelper *helper;
+
+gboolean exit_loop(gpointer user_data)
```
Nit: this should probably be static.

robert-ancell commented on lib/packagekit-glib2/pk-client.c:
```diff
@@ -1965,15 +1964,20 @@ pk_client_create_helper_socket (PkClientState *state)
+	if (g_file_test(socket_filename, G_FILE_TEST_EXISTS))
+		goto out;
+
+	g_free(socket_filename);
```
This is a bit dangerous - socket_filename is now pointing to invalid memory. You should make use a new variable name for the debconf-socket string to avoid a bug being introduced in the future. The other option would be to use `g_clear_pointer (&socket_filename, g_free)` to set it to NULL (I still think multiple variables is clearer and safer).

robert-ancell commented on backends/aptcc/pk-debconf-helper.c:
```diff
@@ -0,0 +1,54 @@
+
+gboolean exit_loop(gpointer user_data)
+{
+	g_debug("Checking for active connections");
```
This file doesn't match the existing code in having space between the function name and parenthesis.

julian-klode replied:
It does match the rest (or most, I did not check it all) of the aptcc backend style, though.

robert-ancell commented:
I don't know enough detail about the intended behaviour with KDE but the code looks otherwise fine aside from the small notes I made.

hughsie commented:
@julian-klode can you fix up the nits please, then it LGTM. I'm planning a release for next week so if it gets merged before then it'll be in the tarball. Thanks.

julian-klode commented:
At least not today, I'm afraid. Probably tomorrow.

julian-klode commented:
Left to do:

- [ ] Rename debconf socket to pk-debconf
- [ ] Add back support for debconf-kde
- [ ] Do not use socket-activated helper when running in terminal or under KDE

julian-klode commented:
OK, should be fine now. I cleaned things up a bit, and fixed a few style issues in the lib/ code. I did not change the style of the pk-debconf-helper in the aptcc backend directory to match the lib/ style, as it currently matches the aptcc style.

julian-klode commented on lib/packagekit-glib2/pk-client-helper.c:
```diff
@@ -487,6 +483,50 @@ pk_client_helper_start (PkClientHelper *client_helper,
+	g_return_val_if_fail (error == NULL || *error == NULL, FALSE);
+
+	/* make sure not been started before */
+	g_return_val_if_fail (priv->argv == NULL, FALSE);
```
Just copy/paste out of `pk_client_helper_start`. And no, it's not normal to call twice, it's only allowed to be called once, as we can only track state of one running client helper (and only need one).

julian-klode commented on lib/packagekit-glib2/pk-client-helper.c:
```diff
@@ -487,6 +483,50 @@ pk_client_helper_start (PkClientHelper *client_helper,
+				    GError **error)
+{
+	gint fd;
+	PkClientHelperPrivate *priv = client_helper->priv;
```
Not going to change this. This should be handled in a different pull request as it's done all over the place.

julian-klode replied:
I guess I'll just fix up the rest, submit a pull request for that and rebase this branch on top of that one.

julian-klode replied:
Rebased on top of #324 with that fixed.

julian-klode commented:
All reported issues should be fixed now. This depends on (or well, includes) #324 which fixes up dereferencing client_helper before checking it.

julian-klode commented:
OK, after the refactoring something broke:

Mai 22 12:41:59 jak-t480s pk-debconf-help[8066]: pk_client_helper_start_with_socket: assertion 'priv->argv == NULL' failed
